pyramid_tm/__init__.py

```diff
@@ -52,67 +64,221 @@ def tm_tween(request):
         ):
             return handler(request)
 
-        # Set a flag in the environment to enable the `request.tm` property.
-        request.environ['tm.active'] = True
-
-        manager = getattr(request, 'tm', None)
-        if manager is None: # pragma: no cover (pyramid < 1.4)
-            manager = create_tm(request)
-            request.tm = manager
-        number = attempts
-        if annotate_user:
-            if hasattr(request, 'unauthenticated_userid'):
-                userid = request.unauthenticated_userid
-            else: # pragma no cover (for pyramid < 1.5)
-                from pyramid.security import unauthenticated_userid
-                userid = unauthenticated_userid(request)
-        else:
-            userid = None
-
-        while number:
-            number -= 1
+        # if we are supporting multiple attempts then we must make
+        # make the body seekable in order to re-use it across multiple
+        # attempts. make_body_seekable will copy wsgi.input if
+        # necessary, otherwise it will rewind the copy to position zero
+        if attempts != 1:
+            request.make_body_seekable()
+
+        # hang onto a reference to the original request as it's the thing
+        # used above the tm tween, we can't change that
+        orig_request = request
+        manager = None
+
+        for number in range(attempts):
+            is_last_attempt = (number == attempts - 1)
+
+            # track the attempt info in the environ
+            # try to set it as soon as possible so that it's available
+            # in the request factory and elsewhere if people want it
+            # note: set all of these values here as they are cleared after
+            # each attempt
+            environ = request.environ
```

Comment:
Since we blow away `request`, could this not be `orig_request.environ` instead? Setting it on `orig_request.environ['xxx']` would possibly be useful as well so that things above `pyramid_tm` can still get at the `tm` information stored in the `environ` (in case they are aware of them).

Maybe not worth it?